### PR TITLE
added reduce function over the samples.

### PR DESCRIPTION
**Pull request by DavideTisi**
it takes as input the tensorMap and the names of the samples you want to sum over. for example if sample_name = ["structure"] it sum over all the sample with the same structure if sample_name = ["structure","center"] it sums over all the rows with the same values for the "structure" and the "center"

it works with the gradients too.

### CODE REVIEW

File: python/src/equistore/operations/_dispatch.py [hunk 1]
```diff
@@ -129,3 +129,15 @@ def vstack(arrays):
         return torch.vstack(arrays)
     else:
         raise TypeError(UNKNOWN_ARRAY_TYPE)
+
+
+def index_add(array, shape0, index):
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:134]: This function feels like it is doing the job of two functions: `zero_like` and then `index_add`. How about splitting it in two?

File: python/src/equistore/operations/_dispatch.py [hunk 2]
```diff
@@ -129,3 +129,15 @@ def vstack(arrays):
...
+    if isinstance(array, np.ndarray):
+        result = np.zeros((shape0,) + array.shape[1:], dtype=np.float64)
+        np.add.at(result, index, array)
+    elif isinstance(array, TorchTensor):
+        result = torch.zeros((shape0,) + array.shape[1:], dtype=torch.float64)
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:139]: This should also allow for different dtypes/devices

File: python/src/equistore/operations/reduce.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    :param sample_name: names of samples to sum
+    """
+
+    try:
+        blocks = []
+        for key, block in tensor:
+            blocks.append(_reduction_sample_block(block=block, sample_name=sample_name))
+        return TensorMap(tensor.keys, blocks)
+    except ValueError:
+        print("it is not possible to perform the reduction for all the blocks")
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:99]: Printing when you caught an exception is not a good practise: the user will likely miss the error and can not do anything about it. Why not just let the exception go all the way to the user here?

File: python/tests/operations/reduce.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        self.assertTrue(
+            np.all(
+                np.sum(bl1.values[26:32], axis=0) == reduce_tensor_ps.block(0).values[6]
+            )
+        )
+        self.assertTrue(
+            np.all(
+                np.sum(bl1.values[32:38], axis=0) == reduce_tensor_ps.block(0).values[7]
+            )
+        )
```
Comment (Luthaf) [inline python/tests/operations/reduce.py:67]: I would not test everything one by one, it makes the test harder to read and modify later if we need to. How about testing the first new sample and one in the middle only?

File: python/src/equistore/operations/reduce.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+from . import _dispatch
+
+# TODO maybe add a way to reduce only the rows
+#  with sample[sample_name] = idx
+# for example not summing over all the sample
+# that have the same 'structure'
+# but only those with 'structure' = 0
+
+
+def _reduction_sample_block(block: TensorBlock, sample_name: list) -> TensorBlock:
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:15]:
```suggestion
def _reduction_sample_block(block: TensorBlock, sample_names: list) -> TensorBlock:
```

since there are multiple names

Comment (Luthaf) [reply python/src/equistore/operations/reduce.py:15]: also, the right type should be

```suggestion
def _reduction_sample_block(block: TensorBlock, sample_name: List[str]) -> TensorBlock:
```

File: python/src/equistore/operations/reduce.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    """Create a new :py:class:`TensorBlock` summing the `properties` among
+    the selected `samples`.
+    The output :py:class:`TensorBlocks` have the same components of the input one.
+    :param block: :py:class:`TensorBlock` -> input block
+    :param sample_name: list -> names of samples to sum
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:20]: you do not need to put the types here again, sphinx should be able to extract them from the signature

File: python/src/equistore/operations/reduce.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    for isample in sample_name:
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:22]: `isample` is a weird name for me: I expect stuff with a `i` in the name to be integers. Why not
```suggestion
    for sample_name in sample_names:
```

or 
```suggestion
    for sample in sample_names:
```

File: python/src/equistore/operations/reduce.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    for parameter, gradient in block.gradients():
+        # here I copy because I want to modify the samples array
+        samples = (
+            gradient.samples.view(dtype=np.int32)
+            .reshape(gradient.samples.shape[0], -1)
+            .copy()
+        )
+        for i in range(len(samples)):
+            samples[i, 0] = index[samples[i, 0]]
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:56]: is there a magical numpy function that would do this without a for loop?

File: python/src/equistore/operations/reduce.py (hunk 6 above)
Comment (DavideTisi) [reply python/src/equistore/operations/reduce.py:20]: this function does not show up in the doc. only resolve that acts on TensorMap does. I can put also this public

Comment (Luthaf) [reply python/src/equistore/operations/reduce.py:20]: No, I think this is fine being private. My comment was more about making docstrings consistent in the whole codebase

File: python/tests/operations/reduce.py (hunk 4 above)
Comment (DavideTisi) [reply python/tests/operations/reduce.py]: i left few of them at the beginning, in the middle and at the end

File: python/src/equistore/operations/reduce.py (hunk 8 above)
Comment (DavideTisi) [reply python/src/equistore/operations/reduce.py:60]: it think samples[:, 0] = index[samples[:, 0]] works

File: python/src/equistore/operations/_dispatch.py [hunk 9]
```diff
@@ -129,3 +129,49 @@ def vstack(arrays):
...
+    it is equivalent of torch's:
+
+    output_array.index_add_(0, torch.tensor(index),input_array)
+
+    """
+    if len(index.shape) != 1:
+        raise ValueError("index should be 1D array")
+    if isinstance(input_array, np.ndarray):
+        _check_all_same_type([output_array], np.ndarray)
+        np.add.at(output_array, index, input_array)
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:148]: This should check that index is a 1D array

File: python/src/equistore/operations/_dispatch.py [hunk 10]
```diff
@@ -129,3 +129,49 @@ def vstack(arrays):
...
+    elif isinstance(input_array, TorchTensor):
+        _check_all_same_type([output_array], TorchTensor)
+        output_array.index_add_(0, torch.tensor(index), input_array)
+    else:
+        raise TypeError(UNKNOWN_ARRAY_TYPE)
+
+
+def zeros_like(array, shape=None, dtype=float, requires_grad=False):
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:156]:
```suggestion
def zeros_like(array, shape=None):
```

File: python/src/equistore/operations/_dispatch.py [hunk 11]
```diff
@@ -129,3 +129,49 @@ def vstack(arrays):
...
+    """Create an zeros_like with the same size of array.
+    if shape is not None it overrides the shape of the result.
+
+    It is equivalent of np.zeros_like(array, dtype=dtype, shape=shape).
+    requires_grad is used only in torch"""
+    if isinstance(array, np.ndarray):
+        result = np.zeros_like(array, dtype=dtype, shape=shape)
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:163]:
```suggestion
        return np.zeros_like(array, dtype=dtype, shape=shape)
```

File: python/src/equistore/operations/_dispatch.py [hunk 12]
```diff
@@ -129,3 +129,49 @@ def vstack(arrays):
...
+    elif isinstance(array, TorchTensor):
+        if shape is None:
+            shape = array.size()
+
+        result = torch.zeros(
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:168]:
```suggestion
        return torch.zeros(
```

File: python/src/equistore/operations/_dispatch.py [hunk 13]
```diff
@@ -129,3 +129,49 @@ def vstack(arrays):
...
+            shape,
+            dtype=array.dtype,
+            layout=array.layout,
+            device=array.device,
+            requires_grad=requires_grad,
+        )
+    else:
+        raise TypeError(UNKNOWN_ARRAY_TYPE)
+    return result
```
Comment (Luthaf) [inline python/src/equistore/operations/_dispatch.py:177]:
```suggestion
```

File: python/src/equistore/operations/reduce.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+from typing import List
+
+# TODO maybe add a way to reduce only the rows
+#  with sample[sample_name] = idx
+# for example not summing over all the sample
+# that have the same 'structure'
+# but only those with 'structure' = 0
+
+
+def _reduction_sample_block(block: TensorBlock, sample_name: List[str]) -> TensorBlock:
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:16]:
```suggestion
def _reduction_sample_block(block: TensorBlock, sample_names: List[str]) -> TensorBlock:
```

File: python/src/equistore/operations/reduce.py [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+    For example if sample_name = ["structure"] it sum over
+    all the sample with the same structure
+    if sample_name = ["structure", "center"] it sums over all the rows with
+    the same values for the "structure" and the "center"
+
+    :param tensor:  input :py:class:`TensorMap`
+    :param sample_name: names of samples to sum
+    """
+
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce.py:97]: We could also check the sample names here using `tensor.sample_names`, and make the check in `_reduction_sample_block` an assertion.

File: python/src/equistore/operations/sum_samples.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+
+    :param tensor:  input :py:class:`TensorMap`
+    :param sample_names: names of samples to sum
+    """
+
+    blocks = []
+    for key, block in tensor:
+        for sample in sample_names:
+            if sample not in block.samples.names:
+                raise ValueError(f"TensorBlock with key={key} has not sample={sample}")
```
Comment (Luthaf) [inline python/src/equistore/operations/sum_samples.py:92]: Since the sample names are the same for all blocks, you can move this check outside of the block loop

File: python/tests/operations/sum_samples.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+import equistore.io
+import numpy as np
+import equistore.operations as fn
+from equistore import Labels, TensorBlock, TensorMap
+
+
+DATA_ROOT = os.path.join(os.path.dirname(__file__), "..", "data")
+
+
+class TestSum_samples(unittest.TestCase):
```
Comment (Luthaf) [inline python/tests/operations/sum_samples.py:13]:
```suggestion
class TestSumSamples(unittest.TestCase):
```

File: docs/src/reference/python/operations/sum_samples.rst [hunk 18]
```diff
@@ -0,0 +1,4 @@
+lstsq
+======
```
Comment (Luthaf) [inline docs/src/reference/python/operations/sum_samples.rst:2]:
```suggestion
sum_samples
===========
```

File: python/src/equistore/operations/sum_samples.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    if sample_name = ["structure", "center"] it sums over all the rows with
+    the same values for the "structure" and the "center"
+
+    :param tensor:  input :py:class:`TensorMap`
+    :param sample_names: names of samples to sum
+    """
+    # I check that the names in sample_names are indeed presents in the block sample
+    # I check only the first one because all the block have the same sample names
+    for sample in sample_names:
+        if sample not in tensor[0].samples.names:
```
Comment (Luthaf) [inline python/src/equistore/operations/sum_samples.py:92]:
```suggestion
    # We check that the names in sample_names are indeed presents in the block sample
    for sample in sample_names:
        if sample not in tensor.sample_names:
```

File: python/src/equistore/operations/sum_samples.py [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+def sum_samples(tensor: TensorMap, sample_names: List[str]) -> TensorMap:
+    """Create a new :py:class:`TensorMap` with the same keys as
+    as the input `tensor`, and each :py:class:`TensorBlock`
+    is obtained summing the corresponding input :py:class:`TensorBlock`
+    over the rows with the same sample_names.
+
+    For example if sample_name = ["structure"], the function sums over
+    all the sample with the same structure
+    if sample_name = ["structure", "center"] it sums over all the rows with
+    the same values for the "structure" and the "center"
```
Comment (Luthaf) [inline python/src/equistore/operations/sum_samples.py:84]:
```suggestion
    For example if ``sample_name = ["structure"]``, the function sums over all
    the sample with the same structure, and if ``sample_name = ["structure",
    "center"]`` it sums over all the rows with the same values for both
    ``"structure"`` and ``"center"`` samples.
```

File: python/src/equistore/operations/sum_samples.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    The output :py:class:`TensorBlocks` have the same components of the input one.
+    :param block: -> input block
+    :param sample_names: -> names of samples to sum
+    """
+    for sample in sample_names:
+        assert (
+            sample in block.samples.names
+        ), "the values in sample_names should be contained in block.samples.names"
+
+    # get the indeces of the selected sample
```
Comment (Luthaf) [inline python/src/equistore/operations/sum_samples.py:22]:
```suggestion
    # get the indices of the selected sample
```

**Review by Luthaf — COMMENTED** [review]
I think this is good to go. Do you want to add the option to use "mean" instead of  "sum" now or leave for later?

File: docs/src/reference/python/operations/sum_samples.rst (hunk 18 above)
Comment (DavideTisi) [reply docs/src/reference/python/operations/sum_samples.rst:2]: I was thinking of doing now, I thought it would not take much but I couldn't think an immediate way to change torch._index _add for the computation of mean. 
for numpy I think you can simply do np.mean.at

**Comment by DavideTisi** [comment]
> I think this is good to go. Do you want to add the option to use "mean" instead of "sum" now or leave for later?

I was thinking of doing now, I thought it would not take much but I couldn't think an immediate way to change torch._index _add for the computation of mean. 
for numpy I think you can simply do np.mean.at

File: python/src/equistore/operations/reduce_over_samples.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    sample_selected = [block.samples.names.index(sample) for sample in sample_names]
+    # reshaping the samples in a 2D array
+    samples = block.samples.view(dtype=np.int32).reshape(block.samples.shape[0], -1)
+    s, index = np.unique(samples[:, sample_selected], return_inverse=True, axis=0)
+    values_result = _dispatch.zeros_like(
+        block.values, shape=(s.shape[0],) + block.values.shape[1:]
+    )
+
+    if reduction == "mean":
+        bincount = _dispatch.bincount(index)
```
Comment (DavideTisi) [inline python/src/equistore/operations/reduce_over_samples.py:43]: to take care of the mean in a way that works both with numpy and torch I used this bitcount thing.
I do not know if it is the best but there are no python loop

File: docs/src/reference/python/operations/sum_over_samples.rst [hunk 23]
```diff
@@ -0,0 +1,4 @@
+sum_over_samples
+================
+
+.. autofunction:: equistore.operations.reduce_over_samples.sum_over_samples
```
Comment (Luthaf) [inline docs/src/reference/python/operations/sum_over_samples.rst:4]:
```suggestion
.. autofunction:: equistore.operations.sum_over_samples
```

File: python/src/equistore/operations/reduce_over_samples.py (hunk 22 above)
Comment (Luthaf) [reply python/src/equistore/operations/reduce_over_samples.py:43]: that looks good!

**Comment by Luthaf** [comment]
I rebased & force-pushed two small cleanup commit. Let me know if everything works for you & then we can merge this!
